### PR TITLE
rgw: add radosclient finisher to perf counter

### PR DESCRIPTION
**Pull request by dongbula**
Signed-off-by: lvshuhua <lvshuhua@cmss.chinamobile.com>

### CODE REVIEW

**Comment by dongbula** [comment]
@cbodley please help review this, thank you :)

**Comment by cbodley** [comment]
hi @dongbula, looks okay to me. but this is a change that could effect the performance of all rados clients, not just radosgw. are you seeing problems in the rados client that require this instrumentation?

**Comment by dongbula** [comment]
@cbodley hi, I just want to add that to perf counter in radosgw, because I find it's difficulty to estimate the proportion of time spent in finisher thread. Yes, you are right, it should not to change all rados client. May be it is good way to add a new constructor for rados client?

`
librados::RadosClient::RadosClient(CephContext *cct_, const string& _name = "");

librados::RadosClient::RadosClient(CephContext *cct_, const string &_name)
  : Dispatcher(cct_->get()),
    cct_deleter{cct_, [](CephContext *p) {p->put();}},
    conf(cct_->_conf),
    state(DISCONNECTED),
    monclient(cct_),
    mgrclient(cct_, nullptr),
    messenger(NULL),
    instance_id(0),
    objecter(NULL),
    lock("librados::RadosClient::lock"),
    timer(cct, lock),
    refcnt(1),
    log_last_version(0), log_cb(NULL), log_cb_arg(NULL)
{
  if (_name == "") {
    finisher = Finisher(cct, _name, "fn" + _name);
  }
  else {
    finisher = Finisher(cct);
  }
}`

**Comment by dongbula** [comment]
@cbodley hi, I want to estimate the proportion of time spent by radosclient finisher thread(radosgw) under high loads, that will be difficulty without perf counter. Does it really have an strong effect on performance?

**Comment by cbodley** [comment]
@liewegas what is your intuition here? is there any risk to adding perf counters to the librados client finisher?

@dongbula i don't think this warrants a change to the librados api (breaking compatibility with existing applications). i think we'll just want to enable it for all librados clients, instead of making it optional

File: src/include/rados/librados.h [hunk 1]
```diff
@@ -406,7 +406,8 @@ CEPH_RADOS_API int rados_create2(rados_t *pcluster,
  * @returns 0 on success, negative error code on failure
  */
 CEPH_RADOS_API int rados_create_with_context(rados_t *cluster,
-                                             rados_config_t cct);
+                                             rados_config_t cct,
+                                             const char *name);
```
Comment (liewegas) [inline src/include/rados/librados.h]: we can't change the ABI here; need to define a new variant of this function to add the arg.

File: src/include/rados/librados.hpp [hunk 2]
```diff
@@ -1221,7 +1221,7 @@ namespace librados
     int init(const char * const id);
     int init2(const char * const name, const char * const clustername,
 	      uint64_t flags);
-    int init_with_context(config_t cct_);
+    int init_with_context(config_t cct_, const std::string& name = "");
```
Comment (liewegas) [inline src/include/rados/librados.hpp]: same here.  it has to be init_with_context_name or something.. don't overload the same name as gonig from non-overloaded to overloaded breaks the C++ ABI

**Comment by liewegas** [comment]
I don't see where the perfcounter comes in here.  Does Finisher take a different code path simply because it has a name or something?

I have some general concerns about perfcounter overhead, but I think they're orthogonal (and likely fixable).  Still, I would be careful about adding counters to very hot paths (e.g., mutex lock/unlock).  I don't see that here though?

**Comment by liewegas** [comment]
Ah. Yeah, let's do it unconditionally.

**Comment by dongbula** [comment]
@liewegas @cbodley 
Good!

**Comment by cbodley** [comment]
@liewegas this passed the [rgw suite](http://pulpito.ceph.com/cbodley-2017-03-17_11:36:22-rgw-wip-cbodley-testing---basic-mira/), does it need to run against others?

**Pull request merged**